actors/builtin/miner/expiration_queue.go

```diff
@@ -698,3 +712,122 @@ func groupSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChai
 	})
 	return sectorEpochSets
 }
+
+// Groups sectors into sets based on their Expiration field.
+// If sectors are not found in the expiration set corresponding to their expiration field
+// (i.e. they have been rescheduled) traverse expiration sets to for groups where these
+// sectors actually expire.
+// Groups will be returned in expiration order, earliest first.
+func (q *ExpirationQueue) findSectorsByExpiration(sectorSize abi.SectorSize, sectors []*SectorOnChainInfo) ([]sectorEpochSet, error) {
+	sectorsByExpiration := make(map[abi.ChainEpoch][]*SectorOnChainInfo)
+
+	for _, sector := range sectors {
+		qExpiration := q.quant.QuantizeUp(sector.Expiration)
+		sectorsByExpiration[qExpiration] = append(sectorsByExpiration[qExpiration], sector)
+	}
+
+	sectorEpochSets := make([]sectorEpochSet, 0, len(sectorsByExpiration))
+	allRescheduled := bitfield.New()
+
+	// This map iteration is non-deterministic but safe because we sort by epoch below.
+	for expiration, epochSectors := range sectorsByExpiration { //nolint:nomaprange // result is subsequently sorted
+		var es ExpirationSet
+		if err := q.mustGet(expiration, &es); err != nil {
+			return nil, err
+		}
+
+		sectorNumbers := []uint64{}
+		totalPower := NewPowerPairZero()
+		totalPledge := big.Zero()
+		for _, sector := range epochSectors {
+			if found, err := es.OnTimeSectors.IsSet(uint64(sector.SectorNumber)); err != nil {
+				return nil, err
+			} else if found {
+				sectorNumbers = append(sectorNumbers, uint64(sector.SectorNumber))
+				totalPower = totalPower.Add(PowerForSector(sectorSize, sector))
+				totalPledge = big.Add(totalPledge, sector.InitialPledge)
+			} else {
+				allRescheduled.Set(uint64(sector.SectorNumber))
+			}
+		}
+		if len(sectorNumbers) > 0 {
+			sectorEpochSets = append(sectorEpochSets, sectorEpochSet{
+				epoch:   expiration,
+				sectors: sectorNumbers,
+				power:   totalPower,
+				pledge:  totalPledge,
+				es:      &es,
+			})
+		}
+	}
+
+	// if sectors are missed because they have been rescheduled, traverse expiration sets to find and group them.
+	if empty, err := allRescheduled.IsEmpty(); err != nil {
+		return nil, err
+	} else if !empty {
+		err := q.traverse(func(epoch abi.ChainEpoch, es *ExpirationSet) (bool, error) {
+			intersect, err := bitfield.IntersectBitField(es.OnTimeSectors, allRescheduled)
+			if err != nil {
+				return false, err
+			}
+
+			group, err := groupSectorSet(sectorSize, sectors, &intersect, es, epoch)
+			if err != nil {
+				return false, err
+			}
+			sectorEpochSets = append(sectorEpochSets, group)
```

Comment:
What if there's already a sectorEpochSet for this epoch, because some other sector had a declared expiration there? I think this will produce another ses for the epoch.

It might be that callers are ok with this, but it's a bit of a trap. Also the `sort.Slice` below is unstable so non-deterministic with equal keys.

I think best to accumulate these in a map by epoch instead, and handle the possibility of the ses already existing at this point.

Reply:
Good catch. I rework the algorithm altogether. It now just intersects `es.OnTimeSectors` with all the sectors it hasn't found yet to form the group and then computes the remainder. It iterates through the expected iterations first and then goes sequentially for what is left. I think it's a little more straightforward, and fixes this issue.